Add recover-tpm package

shjala opened the pull request:
This PR adds the `recovertpm` package. With sufficient knowledge of how EVE intercats with TPM (key indexes, PCR indexes, etc) we can leverage the tool to perform various test on the TPM, re-generate keys, extract the volume key and more.

eriknordmark commented on pkg/mkverification-raw-efi/verify:
```diff
@@ -416,6 +416,14 @@ if [ "$ARCH" = "x86_64" ]; then
+# verify the TPM
+if [ -c $TPM_DEVICE_PATH ]; then
+   logmsg "TPM device is present, running some extra tests"
+   /verifytpm.sh | tee -a "$REPORT/tpmchecks.log" >/dev/console 2>&1
```
If there is a failure, do we indicate that in the summary.log?

shjala replied:
nope, I'll add it.

eriknordmark commented on docs/DEBUGGING.md:
```diff
@@ -451,3 +451,7 @@ Further parsing and diffing the files is possible using tpm2-tools:
+
+### TPM Recovery
+
+Using `recovertpm` tool from debug container, it is possible to perform TPM hardware test, key re-generation, volume key extraction and more. Note that the extracted key is in encrypted wire format suitable for a cloud controler, and key extraction is only possible if the device state (current PCR values) match the original state where key was seald.
```
typo "seald".

Does recovertpm -h list the set of options?
If so it might make sense to mention that here.

shjala replied:
yes it does, you mean just mention `recovertpm -h` or the output of `recovertpm -h` ?

eriknordmark replied:
Just mention it as the way to find out the options.

shjala replied:
done.

eriknordmark reviewed (commented):
Some nits. 
Will re-run yetus workflow since it appears to have failed without any output.

christoph-zededa commented on pkg/debug/recover-tpm/main.go:
```diff
@@ -0,0 +1,1159 @@
+	return cert.PublicKey, nil
+}
+
+func getVualtKeyWireFormat(encryptedVaultKey []byte, digest256 []byte) (string, error) {
```
```suggestion
func getVaultKeyWireFormat(encryptedVaultKey []byte, digest256 []byte) (string, error) {
```

rene commented:
nice tool @shjala , I think it deserves a dedicated repository under lf-edge or under your GH space (or maybe a pkg/recovertpm), then it can be integrated to debugger container through Dockerfile. In this way it can be more easy to maintain, track changes, without mix with debugger container stuff...

eriknordmark reviewed (approved):
LGTM
It seems to be overkill to have a separate repo for this. Mayve a pkg/recovertpm would suffice @rene ?

rene commented:
> LGTM It seems to be overkill to have a separate repo for this. Mayve a pkg/recovertpm would suffice @rene ?

I agree @eriknordmark , pkg/recovertpm should be enough to insulate and track changes....

rouming commented on pkg/debug/recover-tpm/main.go:
```diff
@@ -466,7 +466,7 @@ func main() {
 		}
 
-		encVkeyWire, err := getVualtKeyWireFormat(encVkey, digest256)
+		encVkeyWire, err := getVaultKeyWireFormat(encVkey, digest256)
```
mind to squash this commit ?

shjala replied:
yes, I will move it to pkg/recovertpm and take care of it.

eriknordmark commented:
@rouming there was a request to move the code to pkg/recovertpm instead of it being burried in pkg/debug.

shjala commented:
I will clean up/squash/etc after finished with moving to the a new package.

shjala commented:
it seems lf-edge [JIRA](https://jira.linuxfoundation.org/plugins/servlet/theme/portal/2/group/18) is down (to request a new package on docker-hub).

eriknordmark commented:
> it seems lf-edge [JIRA](https://jira.linuxfoundation.org/plugins/servlet/theme/portal/2/group/18) giving 502 (to request a new package on docker-hub).

I can't read any ticket created by someone else. Are you sure you are logged in as yourself?

Also, is this still WIP? If not, change the title.

shjala commented:
> I can't read any ticket created by someone else. Are you sure you are logged in as yourself?

I can't open the page even, I get this :
<img width="947" alt="image" src="https://github.com/lf-edge/eve/assets/121871672/20858ec7-ae0c-4719-b409-26c23be95826">

Is the URL correct? copied from [BUILD.md](https://github.com/lf-edge/eve/blob/master/docs/BUILD.md#building-packages).

shjala commented on pkg/debug/Dockerfile:
```diff
@@ -8,14 +8,15 @@
 # into the package: see abuild/etc/abuild.conf.
-FROM lfedge/eve-alpine:1f7685f95a475c6bbe682f0b976f12180b6c8726 as build
-ENV BUILD_PKGS abuild curl tar make linux-headers patch g++ git gcc ncurses-dev autoconf
+FROM lfedge/eve-alpine:cbf02c2c126f210933ec9bdb142eb080b400fd76 as build
```
I rebased to resolve the conflict, is the new tag problematic?

eriknordmark replied:
1f7685f95a475c6bbe682f0b976f12180b6c8726 is the tag used in master. So I think you have an old tag

eriknordmark commented:
Yetus is failing to do anything with this in the logs:
Bootstrap
  ./out has been created
  Modes: GitHubActions Robot InContainer ResetRepo UnitTests
  Processing: GH:3867
  GITHUB PR #3867 is being downloaded from
  https://api.github.com/repos/lf-edge/eve/pulls/3867
    JSON data at Wed Jun  5 01:46:14 PM UTC 2024
    Patch data at Wed Jun  5 01:46:14 PM UTC 2024
  ERROR: Unsure how to process GH:3867. Permissions missing?

Any idea why?

shjala commented:
> Yetus is failing to do anything with this in the logs

That is odd, just pushed a new commit and it failed again, looking into it.

shjala commented:
@dautovri suggested to close this one and open a new one to fix the Yetus permission issue. But that did not helped.

shjala commented:
* Squashed and cleaned commits, this one only contains recover-tpm package, rest will follow in another commit.
* Repo on docker hub is ready.
* Rebased to master.
* Fixed yetus complains locally.

eriknordmark reviewed (approved):
LGTM
